Applications/Spire/Include/Spire/Ui/ArrayTableModel.hpp

```diff
@@ -71,10 +72,25 @@ namespace Spire {
 
       boost::signals2::connection connect_operation_signal(
         const typename OperationSignal::slot_type& slot) const override;
+
+    private:
+      mutable OperationSignal m_operation_signal;
+      std::vector<std::vector<std::any>> m_data;
+      Transaction m_transaction;
+      int m_transaction_level;
   };
 
   template<typename F>
   decltype(auto) ArrayTableModel::transact(F&& transaction) {
+    m_transaction_level++;
+    BOOST_SCOPE_EXIT(this_) {
+      this_->m_transaction_level--;
+      if(this_->m_transaction_level == 0) {
+        this_->m_operation_signal(this_->m_transaction);
```

Comment:
Nice solution overall, this part isn't reentrant however. You will need to do something like:

    auto transaction = std::move(this_->m_transaction);
    this_->m_transaction.clear();
    this_->m_operation_signal(transaction);

Reply:
I don't understand why we should do that. The basic idea of the solution is all operations during a transaction and its nested transactions are stored in m_transaction for one call of transact, and when all transactions including child transactions are complete ( m_transaction_level == 0), the transaction recording all operations will be signalled and the m_transaction will be cleared for the next call of the function transact.

Reply:
Because this implementation isn't reentrant. Re-entrancy is a very tricky property to implement correctly and will introduce bugs. Add a unit test to ensure that the solution is re-entrant and then feel free to provide a solution for it. A re-entrancy test would be something along the lines of this pseudocode:

    table = ArrayTableModel();
    table.connect_signal([&] (const Transaction& transaction) {
      table.push({7, 8, 9});
      REQUIRE(transaction.m_operations == 2);
    }
    table.transact([&] {
      table.add({1, 2, 3});
      table.add({4, 5, 6});
    });